drivers/console/semihosting_console.c

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2020 Maciej Sopyło <maciek134@gmail.com>
+ * Copyright (c) 2018 Josef Gajdusek <atx@atx.name>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+#include <kernel.h>
+#include <sys/printk.h>
+#include <device.h>
+#include <init.h>
+#include <string.h>
+
+extern void __printk_hook_install(int (*fn)(int));
+extern void __stdout_hook_install(int (*fn)(int));
+
+static char buffer[CONFIG_SEMIHOSTING_BUFFER_SIZE];
+static u32_t buffer_pos;
+
+static void semihosting_flush_buffer(struct k_timer *timer);
+static K_TIMER_DEFINE(send_timer, semihosting_flush_buffer, NULL);
+
+
+static void semihosting_flush_buffer(struct k_timer *timer)
+{
+	if (buffer[buffer_pos] != 0) {
+		buffer[buffer_pos++] = 0;
+	}
+
+	register int r0 __asm__("r0") = 0x04; // SYS_WRITE0
+	register char *r1 __asm__("r1") = buffer;
+	__asm__ volatile("bkpt $0xab"
+			:
+			: "r" (r0), "r" (r1));
+
+	buffer_pos = 0;
+}
+
+
+static int semihosting_console_out(int character)
+{
+	int key = irq_lock();
+	bool end;
+
+	buffer[buffer_pos++] = character;
+	// leave space for the trailing zero
+	end = buffer_pos >= ARRAY_SIZE(buffer) - 1;
+
+	/* Restart timer if it expired */
+	if (k_timer_remaining_get(&send_timer) == 0) {
+		k_timer_start(&send_timer,
+			K_MSEC(CONFIG_SEMIHOSTING_SEND_TIMEOUT_MS),
+			Z_TIMEOUT_TICKS(0));
+	}
+
+	irq_unlock(key);
+
+	if (end) {
+		k_timer_status_sync(&send_timer);
+	}
```

Comment:
I am not convinced this is necessary.

`k_timer_start` can be called on an already running timer, so the char output function can blindly call `k_timer_start` with some single shot timeout to trigger a flush when no character output is requested for the timeout interval.